### DISCUSSION TITLE
카테고리에 Root 카테고리가 필요할까요?

### DISCUSSION THREAD

**Discussion by YeaChan05** [KakaoFunding/back-end#26] [Q&A] [answered]
지난번  pr때(#13 ) 카테고리를 자기참조와 계층화를 통해 관리하기 위해 로직을 작성했었습니다
지난 pr 코드를 보시면 아시겠지만 카테고리의 가장 최상단에는 DTO계층에서는 아래와 같이 ROOT dto를 만들어 모든 카테고리를 계층화하여 응답 가능하도록 로직을 작성했었습니다
```Java
private static final CategoryDto ROOT_CATEGORY_DTO = new CategoryDto(-1L, "Root", null);
```

하지만 이 방법은 모든 카테고리를 전부 조회해온 후에 application단에서 전부 매핑을 해주는 방식이라 성능도 좋지 못하고 부모 계층의 부모 카테고리를 null로 비워두어야한다는 단점이 있었습니다

그래서 카테고리 entity단에도 root 카테고리를 넣는것이 어떤가에 대한 의견 여쭙고자합니다!

**Comment by kmw2378** [comment]
엔티티단에 root 카테고리를 넣는다는게 무슨 말씀인지 이해가 안가네요ㅜㅜㅜ `Category` 엔티티 필드에 `root` 필드를 추가한다는 말씀이실까요??

**Reply by YeaChan05** [reply]
기존 로직은 최상위 카테고리인 root 카테고리가 db에 존재하지 않는 로직을 작성했었습니다
최상위 카테고리의 경우에는 카테고리의 부모가 null이어야하는 경우가 발생해서 root에대한 카테고리만 매번 테스트할때 번거로워서 빼고 생각했었는데 결국 root 카테고리를 가져오고, fetch join으로 한번에 가져오는게 차라리 낫겠다라는 판단이 들더라고요

**Reply by YeaChan05** [reply]
방금 이방법으로 해결해보려 헀는데 fetch join 만으로는 root에서 최하단 카테고리까지 조회하기에는 `MultipleBagFetchException`이 발생할수밖에 없긴 하네요... 조회성능이 좀 안좋긴 해도 카테고리는 변경이 매우 적은 도메인이라 캐싱해두고 변경시에만 수정해서 사용하는 방법을 쓰면 좋지 않을까 싶네요

**Reply by kmw2378** [reply, marked as answer]
말씀하신대로 DB단에서 튜닝을 해도 한계가 있을 것 같아요! 캐싱을 하신다면 어떻게 하실 생각이실까요?

**Reply by kmw2378** [reply]
그리고 제가 카테고리를 가져오는 로직이 자세히 이해가 안되서 혹시 시간나실 때 디코로 설명해주실 수 있을까요??
